File: src/game/plot/intro.ts
// === Plot Module: Intro ===
//
// Lore slideshow (intro0–intro8) → sex choice → realm choice.
// Pure content — if this file is deleted the game runs without an intro.

import {EventTag, type StoryPhase} from '../event-types';
import {type LogicNode, createNode} from '../logic-nodes';

// ── Slide narration table ──

const SLIDES: StoryPhase = {
	type: 'slides',
	slides: [
		{image: '/assets/backgrounds/intro0.png', narration: 'In the time before memory, the gods shaped a world upon the surface of a torus — infinite yet bounded.'},
		{image: '/assets/backgrounds/intro1.png', narration: 'Pure Magic flowed through every stone and river, the breath of creation itself.'},
		{image: '/assets/backgrounds/intro2.png', narration: 'But the gods grew jealous of their own work… and destroyed one another.'},
		{image: '/assets/backgrounds/intro3.png', narration: 'Their corpses became the Black Force — void and negation, whispering from beyond.'},
		{image: '/assets/backgrounds/intro4.png', narration: 'Where Pure Magic and Black Force meet, both are annihilated. The world trembles.'},
		{image: '/assets/backgrounds/intro5.png', narration: 'Kingdoms rose. Mage-lords built towers of arrogance. Empires banned magic under pain of death.'},
		{image: '/assets/backgrounds/intro6.png', narration: 'Barbarian kings seized castles from slain wizards, promising freedom they could not deliver.'},
		{image: '/assets/backgrounds/intro7.png', narration: 'A prophecy speaks of a Black Child — herald of the end of Pure Magic.'},
		{image: '/assets/backgrounds/intro8.png', narration: 'And now, traveller, you arrive. The world does not yet know your name.'},
	],
};

// ── Choice phases ──

const SEX_CHOICE: StoryPhase = {
	type: 'choice',
	id: 'sex',
	title: 'Choose Your Nature',
	description: 'The body shapes the mind, and the mind shapes destiny.',
	options: [
		{
			label: 'Male', description: 'Strong mind — +1 skill point', value: 'male', image: '/assets/sprites/male.png',
		},
		{
			label: 'Female', description: 'Strong body — +1 attribute point', value: 'female', image: '/assets/sprites/female.png',
		},
	],
};

const NAME_INPUT: StoryPhase = {
	type: 'input',
	id: 'name',
	title: 'What Is Your Name?',
	description: 'A name is the first claim a soul makes upon the world.',
	placeholder: 'Enter your name',
	defaultValue: 'Traveller',
	maxLength: 24,
};

const REALM_CHOICE: StoryPhase = {
	type: 'choice',
	id: 'realm',
	title: 'Choose Your Homeland',
	description: 'Where you were born determines who you must become — or defy.',
	options: [
		{label: 'Magocracy of Magika', description: 'Land of mage-lords. Magic is everyday life.', value: 'magika'},
		{label: 'Empire of Light', description: 'Holy empire. Magic is forbidden on pain of death.', value: 'empire'},
		{label: 'Barbarian Kingdoms', description: 'Feudal warlords ruling by sword and steel.', value: 'barbarians'},
	],
};

// ── Nodes ──

export const introNodes: LogicNode[] = [
	createNode({
		id: 'intro_main',
		label: 'Intro Sequence',
		conditions: [],
		mask: [],
		next: [],
		tags: ['intro', 'plot'],
		effect(ctx) {
			ctx.bus.emit({
				tag: EventTag.ShowStory,
				phases: [SLIDES, SEX_CHOICE, NAME_INPUT, REALM_CHOICE],
				sourceNodeId: 'intro_main',
			});
		},
	}),
];

export const introActiveNodes = ['intro_main'];

File: timaert_c/src/content/plot/intro.cpp
#include "content/plot/intro.h"
#include "events/event_types.h"
#include "events/logic_nodes.h"

#include <cstdint>

namespace sm::content {
namespace {

constexpr StorySlide kIntroSlides[] = {
    {"/assets/backgrounds/intro0.png", "In the time before memory, the gods shaped a world upon the surface of a torus - infinite yet bounded."},
    {"/assets/backgrounds/intro1.png", "Pure Magic flowed through every stone and river, the breath of creation itself."},
    {"/assets/backgrounds/intro2.png", "But the gods grew jealous of their own work... and destroyed one another."},
    {"/assets/backgrounds/intro3.png", "Their corpses became the Black Force - void and negation, whispering from beyond."},
    {"/assets/backgrounds/intro4.png", "Where Pure Magic and Black Force meet, both are annihilated. The world trembles."},
    {"/assets/backgrounds/intro5.png", "Kingdoms rose. Mage-lords built towers of arrogance. Empires banned magic under pain of death."},
    {"/assets/backgrounds/intro6.png", "Barbarian kings seized castles from slain wizards, promising freedom they could not deliver."},
    {"/assets/backgrounds/intro7.png", "A prophecy speaks of a Black Child - herald of the end of Pure Magic."},
    {"/assets/backgrounds/intro8.png", "And now, traveller, you arrive. The world does not yet know your name."},
};

constexpr StoryChoice kSexChoices[] = {
    {"Male", "Strong mind - +1 skill point", "male", "/assets/sprites/male.png"},
    {"Female", "Strong body - +1 attribute point", "female", "/assets/sprites/female.png"},
};

constexpr StoryChoice kRealmChoices[] = {
    {"Magocracy of Magika", "Land of mage-lords. Magic is everyday life.", "magika", nullptr},
    {"Empire of Light", "Holy empire. Magic is forbidden on pain of death.", "empire", nullptr},
    {"Barbarian Kingdoms", "Feudal warlords ruling by sword and steel.", "barbarians", nullptr},
};

constexpr StoryPhaseDef kIntroPhases[] = {
    {
        StoryPhaseKind::Slides,
        nullptr,
        nullptr,
        nullptr,
        kIntroSlides,
        sizeof(kIntroSlides) / sizeof(kIntroSlides[0]),
        nullptr,
        0,
        nullptr,
        nullptr,
        0,
    },
    {
        StoryPhaseKind::Choice,
        "sex",
        "Choose Your Nature",
        "The body shapes the mind, and the mind shapes destiny.",
        nullptr,
        0,
        kSexChoices,
        sizeof(kSexChoices) / sizeof(kSexChoices[0]),
        nullptr,
        nullptr,
        0,
    },
    {
        StoryPhaseKind::Input,
        "name",
        "What Is Your Name?",
        "A name is the first claim a soul makes upon the world.",
        nullptr,
        0,
        nullptr,
        0,
        "Enter your name",
        "Traveller",
        24,
    },
    {
        StoryPhaseKind::Choice,
        "realm",
        "Choose Your Homeland",
        "Where you were born determines who you must become - or defy.",
        nullptr,
        0,
        kRealmChoices,
        sizeof(kRealmChoices) / sizeof(kRealmChoices[0]),
        nullptr,
        nullptr,
        0,
    },
};

constexpr StoryDef kIntroStory = {
    "intro",
    "intro_main",
    kIntroPhases,
    sizeof(kIntroPhases) / sizeof(kIntroPhases[0]),
};

constexpr std::uint32_t total_choices(const StoryDef& story) {
    std::uint32_t total = 0;
    for (std::size_t i = 0; i < story.phaseCount; ++i) {
        total += static_cast<std::uint32_t>(story.phases[i].choiceCount);
    }
    return total;
}

constexpr int input_max_length(const StoryDef& story) {
    for (std::size_t i = 0; i < story.phaseCount; ++i) {
        if (story.phases[i].kind == StoryPhaseKind::Input) {
            return story.phases[i].maxLength;
        }
    }
    return 0;
}

LogicNode intro_main_node() {
    LogicNode n;
    n.id = kIntroStory.sourceNodeId;
    n.label = "Intro Sequence";
    n.tags.push_back("intro");
    n.tags.push_back("plot");
    n.effect = [](NodeContext& ctx) {
        const StoryDef& story = intro_story();
        GameEvent ev{EventTag::ShowStory};
        ev.s1 = story.sourceNodeId;
        ev.s2 = story.id;
        ev.ix = static_cast<int>(story.phaseCount);
        ev.iy = story.phaseCount > 0 ? static_cast<int>(story.phases[0].slideCount) : 0;
        ev.a = total_choices(story);
        ev.b = static_cast<std::uint32_t>(input_max_length(story));
        ctx.bus->emit(ev);
    };
    return n;
}

} // namespace

const StoryDef& intro_story() {
    return kIntroStory;
}

void register_intro_story_nodes(LogicNodeEngine& logic) {
    logic.add(intro_main_node());
}

} // namespace sm::content
